modules/lib/mabAddressbookUtil.js

```diff
@@ -56,5 +59,117 @@ var AddressbookUtil = {
 
       foStream.close();
     }
+  },
+
+  /**
+   * Ask the user to point to a vCard and import the contacts into the database
+   *
+   * @param {Addressbook} Addressbook to add the imported contacts to
+   * @returns {Promise[]} Array of promises for adding the new contacts
+   */
+  importContacts: function(addressbook) {
+
+    var filePickerInterface = Components.interfaces.nsIFilePicker;
+    var filePicker = Components.classes["@mozilla.org/filepicker;1"].createInstance(filePickerInterface);
+    var windowMediator = Cc["@mozilla.org/appshell/window-mediator;1"].getService(Ci.nsIWindowMediator);
+    var window = windowMediator.getMostRecentWindow(null);
+
+    filePicker.init(window, "Load contacts from", filePickerInterface.modeOpen);
+    filePicker.appendFilter("vCard", "*.vcf; *.vcard");
+    filePicker.appendFilters(filePickerInterface.filterAll);
+    filePicker.filterIndex = 0;
+    filePicker.defaultString = "contacts.vcf";
+
+    var returnValue = filePicker.show();
+
+    if (returnValue == filePickerInterface.returnOK) {
+
+      var inputStream = Components.classes["@mozilla.org/network/file-input-stream;1"]
+        .createInstance(Components.interfaces.nsIFileInputStream);
+
+      inputStream.init(filePicker.file, 0x01, 0444, 0); // readonly
+
+      var fileContents = NetUtil.readInputStreamToString(
+          inputStream,
+          inputStream.available(),
+          {});
+
+      inputStream.close();
+
+      var contacts = ICAL.parse(fileContents);
+      // detect error from parsing
+      if (!contacts) {
+        throw "Invalid vCard";
+      }
+
+      // check if it is an array of vcards or a single vcard
+      if (contacts[0] == 'vcard') {
+        // make a single contact a list to work with the next section
+        contacts = [contacts];
+      }
+
+      // convert jcards into Contact objects
+      var contactPromises = contacts.map(function(vcard) {
+        var contact =  new ICAL.Component(vcard);
+
+        // try to get the name of the contact
+        var name = contact.getFirstPropertyValue("fn");
+        // fall back to the 'n' for a name
+        if (!name) {
+          name = contact.getFirstPropertyValue("n");
+          name = Array.isArray(name) ? name.join(" ").trim() : name;
+        }
+        // fall back to email for a name
+        if (!name) {
+          name = contact.getFirstPropertyValue("email");
+        }
+
+        // get the photo
+        var photo = undefined;
+        var photoPropertry = contact.getFirstProperty("photo");
+
+        if (photoPropertry) {
+          if (photoPropertry.type === "binary") {
+            var imageType = photoPropertry.getParameter("type").toLowerCase();
+            photo = AddressbookUtil.b64toBlob(photoPropertry.getValues(), imageType);
+          }
+        }
+
+        // TODO: check if it already exists and add it?
+
+        // add the contact to the addressbook
+        addressbook.add({
```

Comment:
this wont add it to a contact if it already exists right? it will add it as a new contact as we dont have the uuid to find a similar contact. does this mean that we wont deal with merging? let people decide to do it later (tho its not a feature in scope for us, can b for next team)